lib/OpenQA/Scheduler.pm

```diff
@@ -37,68 +30,77 @@ use constant MAX_JOB_ALLOCATION => $ENV{OPENQA_SCHEDULER_MAX_JOB_ALLOCATION} //
+our $RUNNING;
 
-    my $ret = $self->SUPER::_is_method_allowed($method);
-    if ($ret) {
-        log_debug "IPC calling $method";
-    }
-    return $ret;
-}
-
-our $setup;
 sub run {
+    my $self   = __PACKAGE__->new;
+    my $daemon = $self->setup;
+    local $RUNNING = 1;
+    $daemon->run;
+}
 
-    # Catch normal signals
-    local $SIG{HUP} = local $SIG{INT} = local $SIG{PIPE} = local $SIG{TERM} = sub {
-        OpenQA::Scheduler::Scheduler::quit();
-    };
+sub setup {
+    my $self = shift;
 
-    $setup = OpenQA::Setup->new(log_name => 'scheduler');
+    my $setup = OpenQA::Setup->new(log_name => 'scheduler');
     OpenQA::Setup::read_config($setup);
     OpenQA::Setup::setup_log($setup);
 
-    OpenQA::Scheduler->new();
     log_debug("Scheduler started");
     log_debug("\t Scheduler default interval(ms) : " . SCHEDULE_TICK_MS);
-    log_debug("\t Max job allocation: " . MAX_JOB_ALLOCATION);
-
-    my $reactor = Net::DBus::Reactor->main;
-    OpenQA::Scheduler::Scheduler::reactor($reactor);
-    $reactor->{timer}->{schedule_jobs} = $reactor->add_timeout(
-        SCHEDULE_TICK_MS,
-        Net::DBus::Callback->new(
-            method => \&OpenQA::Scheduler::Scheduler::schedule
-        ));
+    log_debug("\t Max job allocation: " . OpenQA::Scheduler::Model::Jobs::MAX_JOB_ALLOCATION());
+
     # initial schedule
-    OpenQA::Scheduler::Scheduler::schedule();
+    OpenQA::Scheduler::Model::Jobs->singleton->schedule;
+    Mojo::IOLoop->next_tick(sub { _reschedule() });
 
-    $reactor->run;
+    return Mojo::Server::Daemon->new(app => $self);
 }
 
-sub new {
-    my ($class) = @_;
-    $class = ref $class || $class;
-    # register @ IPC - we use DBus reactor here for symplicity
-    my $ipc = OpenQA::IPC->ipc;
-    return unless $ipc;
-    my $service = $ipc->register_service('scheduler');
-    my $self    = $class->SUPER::new($service, '/Scheduler');
-    $self->{ipc} = $ipc;
-    bless $self, $class;
-
-    return $self;
+sub startup {
+    my $self = shift;
+
+    $self->defaults(appname => 'openQA Scheduler');
+    $self->mode('production');
+
+    # no cookies for worker, no secrets to protect
+    $self->secrets(['nosecretshere']);
+    $self->config->{no_localhost_auth} ||= 1;
+
+    # The reactor interval might be set to 1 ms in case the scheduler has been woken up by the
+    # web UI (In this case it is important to set it back to OpenQA::Scheduler::SCHEDULE_TICK_MS)
+    OpenQA::Scheduler::Model::Jobs->singleton->on(
+        conclude => sub {
+            _reschedule(SCHEDULE_TICK_MS);
+        });
+
+    my $r = $self->routes;
+    $r->namespaces(['OpenQA::Scheduler::Controller', 'OpenQA::Shared::Controller']);
+    my $ca = $r->under('/')->to('Auth#check');
+    $ca->get('/' => {json => {name => $self->defaults('appname')}});
+    my $api = $ca->any('/api');
+    $api->get('/wakeup')->to('API#wakeup');
+    $r->any('/*whatever' => {whatever => ''})->to(status => 404, text => 'Not found');
```

Comment:
I don't understand what these route handlers(?) are for, is it just for experimentation?

Reply:
It's a clean fallback for the case where no route matches. Since this is purely an API server i like having a succinct text message instead of the default HTML file. The websocket server does the same.